File: SPCode/UI/ComboBox/SPComboBox.vb
Namespace UI
    Public Class SPComboBox
        Inherits ComboBox

        Private m_strLookupText As String = "<New...>"
        Private m_intLookupIndex As Integer

        Public Event Populate(ByVal sender As Object, ByVal evt As SPComboBoxEventArgs)
        Public Event Lookup(ByVal sender As Object, ByVal evt As SPComboBoxEventArgs)


        Public Property LookupText() As String
            Get
                Return m_strLookupText
            End Get
            Set(ByVal value As String)
                m_strLookupText = value
            End Set
        End Property
        Public Property LookupIndex() As Integer
            Get
                Return m_intLookupIndex
            End Get
            Set(ByVal value As Integer)
                m_intLookupIndex = value
            End Set
        End Property

        Protected Overrides Sub OnSelectedValueChanged(ByVal e As System.EventArgs)
            MyBase.OnSelectedValueChanged(e)

            If Me.SelectedIndex = LookupIndex Then
                RaiseEvent Lookup(Me, New SPComboBoxEventArgs)
            End If
        End Sub


    End Class

    Public Class SPComboBoxEventArgs
        Inherits EventArgs

        Private m_objData As Object

        Public Property Data() As Object
            Get
                Return m_objData
            End Get
            Set(ByVal value As Object)
                m_objData = value
            End Set
        End Property
    End Class
End Namespace

File: SPCode/CS/UI/ComboBox/SPComboBox.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;

namespace SoftLogik.Win
{
	namespace UI
	{
		public partial class SPComboBox : ComboBox
		{
			
			
			private string m_strLookupText = "<New...>";
			private int m_intLookupIndex;
			
			public delegate void PopulateEventHandler(object sender, SPComboBoxEventArgs evt);
			private PopulateEventHandler PopulateEvent;
			
			public event PopulateEventHandler Populate
			{
				add
				{
					PopulateEvent = (PopulateEventHandler) System.Delegate.Combine(PopulateEvent, value);
				}
				remove
				{
					PopulateEvent = (PopulateEventHandler) System.Delegate.Remove(PopulateEvent, value);
				}
			}
			
			public delegate void LookupEventHandler(object sender, SPComboBoxEventArgs evt);
			private LookupEventHandler LookupEvent;
			
			public event LookupEventHandler Lookup
			{
				add
				{
					LookupEvent = (LookupEventHandler) System.Delegate.Combine(LookupEvent, value);
				}
				remove
				{
					LookupEvent = (LookupEventHandler) System.Delegate.Remove(LookupEvent, value);
				}
			}
			
			
			
			public string LookupText
			{
				get
				{
					return m_strLookupText;
				}
				set
				{
					m_strLookupText = value;
				}
			}
			public int LookupIndex
			{
				get
				{
					return m_intLookupIndex;
				}
				set
				{
					m_intLookupIndex = value;
				}
			}
			
			protected override void OnSelectedValueChanged(System.EventArgs e)
			{
				base.OnSelectedValueChanged(e);
				
				if (this.SelectedIndex == LookupIndex)
				{
					if (LookupEvent != null)
						LookupEvent(this, new SPComboBoxEventArgs());
				}
			}
			
			
		}
		
		public class SPComboBoxEventArgs : EventArgs
		{
			
			
			private object m_objData;
			
			public object Data
			{
				get
				{
					return m_objData;
				}
				set
				{
					m_objData = value;
				}
			}
		}
	}
	
	
	
	
}
